### ISSUE TITLE
Service call failing when called through HttpClient when fiddler is not running and passes when fiddler is running

### ISSUE THREAD

**Issue by girija45701** [amzn/selling-partner-api-docs#166]
Hi all,

I am trying to use HttpClient instead of restsharp and facing a strange issue calling the Seller api. The strange thing is that if fiddler is on the same code works but fails when fiddler is not running.

When fiddler is on it passes and gets me result. When fiddler is not running it fails. Since it was not failing when fiddler was on, i installed another tool "Http Debugger" to see the failure and i could not make any difference in both the request (The pass and the failure). Bothe of them are given below

Failure Capture:

```
GET /sellers/v1/marketplaceParticipations HTTP/1.1
Accept: application/json
x-amz-access-token: Atza|IwEBIN3hOVtNi1xM47txtHcqXi5A3C960AypB7pkWYCxEo7lNiL9EFR-1b_EoD6PQ8lzAXgM4zujF0OBv0NS7sYQ9bWqPMDhHFt8kgvdmlmk3==jrinknrO6PYlZgmFLRGn1Hzmvgldmnj4973bjkfnbkldlcvld vc0BmuqKHUreonrWQxFO49u0yoIiNHVzSxHP0Wo4nWKW5pdd5Fj73gYxnZQQeYF5EAy8lKDCLtndTnJCdlrv5Kk8JK8iFD_H7h3FF5H4gNyTx3uIHxMaU8OkLz_IigsCTNQHwljnubhQlR9aK0J6lRbb0QfOQ4BAT_e1GOKDkShu-U5OdchdF5qNUkKU
user-agent: MSolution/1.0.0.0
X-Amz-Date: 20201201T224620Z
Authorization: AWS4-HMAC-SHA256 Credential=AKIA5SU6JNBJKDAQB76813QI6V/20201201/eu-west-1/execute-api/aws4_request, SignedHeaders=accept;host;user-agent;x-amz-access-token;x-amz-date, Signature=71dd12ee0eaf33cd142dwr242424e91cb5c4bfd6fd4f46d929d
Host: sellingpartnerapi-eu.amazon.com
Connection: Keep-Alive
```

It fails with Forbidden message

The success capture is (Fiddler running):

```
GET /sellers/v1/marketplaceParticipations HTTP/1.1
Accept: application/json
x-amz-access-token: Atza|IwEBIMfdZDDaca8HrDGIPft-HQs3Vzi75I4Bk9iNKfsHTkfwsfsfcsvcsaP86DqKkoZE37TiDr3XvmD_vdvavcUE9TzdXhf2jjuULL04keBHI_XYrnTnhXaCPE0gUAc8HvIiW7OXSERz_3RlS9R-nu2lTo_jqzaz0mbUaa-evavaVAVLauh2Ue7Io8pE1tThRTcqM60igPcrBViAUptTAsq-IL5ZT7hOfbNJTJ31GeN8e8IzjkWfe9n4l7B799VM1bJnC-D_alZ2J0HHj4cBNjd3RzAEvavav3fGWkW5iH2_MZ3IyaxYnslvSzNH4h8tvay87OywkkxVUKIn
user-agent: MSolution/1.0.0.0
X-Amz-Date: 20201201T224820Z
Authorization: AWS4-HMAC-SHA256 Credential=AKIAYWBDAQ7XYIPQI6V/20201201/eu-west-1/execute-api/aws4_request, SignedHeaders=accept;host;user-agent;x-amz-access-token;x-amz-date, Signature=7e4db1c114219546848eaffnvclknslnlcvs63c9e0af50edc3cdbe7231c9b
Host: sellingpartnerapi-eu.amazon.com
```
I have already used:

```
ServicePointManager.ServerCertificateValidationCallback +=
                  (senderSP, certificate, chain, sslPolicyErrors) => true;

            ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls | SecurityProtocolType.Tls11 | SecurityProtocolType.Tls12 | SecurityProtocolType.Ssl3;
```

I know this question is more on implementation and not the api, but what i am looking is if anyone has faced this. Is there something i am missing that i should take care for teh API when using the HttpClient classes (which RestSharp library may be automatically handling).

Any advise or pointers would be helpful.

- Shankar

**Comment by JacksonJeans** [comment]
Hello again @girija45701 ,

In your request that returns "Forbidden" you send a `Connection` header, but it is not signed in your `Authorization` header.
All headers must appear in the `SignedHeaders`-parameter in the `Authorization`-header if you want to submit a successful request.

Try to omit the `Connection`-header in your implementation or mention the header name in the `SignedHeaders`-parameter.

Please let me know if it worked.

Good luck!

**Comment by girija45701** [comment]
@JacksonJeans ,

Thanks for the reply. Since i am using HttpClient it seems it is not possible to remove the Connection from the same.

I tried adding the Connection in the signed header and the value in the canonical request as well. But it still fails.

Here is my raw request now.

```
GET /sellers/v1/marketplaceParticipations HTTP/1.1
Accept: application/json
x-amz-access-token: Atza|IwEBIBYvkW6rk1NjPExvUvYpVPbnkbkb79jbjb7E5_odf2jJDaxhlnNvYT3yf4_x8U3W7MyZIujdQZYiAb2QBcA0fb9lAKyDDHAx2e5R-_Gv-c_yPtJK8sgn86G2aDt5_sdUtIqhWLqJoVgXAOo6WSKwpyrA2-UmGCc__JRRR-erc1qR4WNEv1pinA8n50o9SU2t1bnlkbnl-bjbbllgiobiu8ghuug8V1kmE7B7-WqXGP4zAk4abdDQ_-oXDiaiq8XDkbL0ngtPnBjLVjPe91v17iqik9bR4BDbVEC1ptUUhDmhJ-YkBqVONV-c_rc76ZtKDLR1R6H5f
user-agent: MSolution/1.0.0.0
X-Amz-Date: 20201202T153418Z
Authorization: AWS4-HMAC-SHA256 Credential=AKIA5SU6XCGCX47CGCGQI6V/20201202/eu-west-1/execute-api/aws4_request, SignedHeaders=accept;connection;host;user-agent;x-amz-access-token;x-amz-date, Signature=7f6gbuj878701ea66fa3141fvjubikj79hibab9f1be7dfa922e6f4
Host: sellingpartnerapi-eu.amazon.com
Connection: Keep-Alive
```
- Shankar

**Comment by JacksonJeans** [comment]
@girija45701 ,
Can you tell me which headers you get back in the Forbidden-message? Is it a SignatureException? What exactly is the error message? 

Otherwise my Latin ends up to here. The canonical request looks good now.

**Comment by girija45701** [comment]
@JacksonJeans,

My response is just "Forbidden". Simply that message. There is no message like "Signature Invalid". This is where I am not a big fan of the the API responses. Does not tell me what exactly is the error.

- Shankar

**Comment by rogersv** [comment]
I have noticed that the SP-API sometimes respond with a 403 with the extra message "Unknown client". It happened for me when I used postman. I think that SP-API did not accept the user agent that was added by postman. Try to change user agent.
Also see https://github.com/amzn/selling-partner-api-docs/blob/main/guides/developer-guide/SellingPartnerApiDeveloperGuide.md#include-a-user-agent-header-in-all-requests where it says that language is a required attribute to the user-agent.
